shinylive/_url.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+import base64
+import json
+import os
+import re
+from pathlib import Path
+from typing import List, Literal, Optional, TypedDict, cast
+
+
+class FileContentJson(TypedDict):
+    name: str
+    content: str
+    type: Literal["text", "binary"]
+
+
+def encode_shinylive_url(
+    app: str | Path,
```

Comment:
I feel uneasy about allowing `app` to be a string which is either a path, or the content of an app.py/app.R file. This kind of dual purpose of a parameter feels weird, and it has bitten us in the past (IIRC, with jsonlite).

I think there should just be a different parameter, like `app_contents`, for that case.

Reply:
In light of also wanting to add an "encode to JSON" option, I think it makes sense to break apart the work being done here in `encode_shinylive_url()` to break it into two steps of creating the `list[FileContentJson]` bundle and to encode the bundle directly into a URL. With that pairing, we could have separate functions that perform each task rather than overloading a single argument.

At that point, we'd be pushing this check onto users to decide which function to call, but that doesn't seem terribly cumbersome.

Reply:
> I feel uneasy about allowing `app` to be a string which is either a path, or the content of an app.py/app.R file. This kind of dual purpose of a parameter feels weird

I agree that the dual purpose `app` parameter feels weird, but I feel equally or slightly more weird about two competing required parameters. If we added `app_contents`, then only one of `app` or `app_contents` would be allowed, since they would both serve the purpose of _the primary app_.

In the latest commit, I've broken this into a two steps:

1. First choose `create_shinylive_bundle_file()` or `create_shinylive_bundle_text()` depending on whether you want to pass `app` as the path to an app **file** or `app` as the **text** contents of an app.
  
   Both functions return an _app bundle_ which is just a dictionary containing the list of `FileContentJson` as `files`, plus the `language` since we might have needed to detect it from the value of `app`.

2. Then pass this bundle to `create_shinylive_url()` where there a few additional options, such as `mode` and `header`.

This certainly feels more composable to me, and I like the idea of being able to give `create_shinylive_url()` a fully specified app bundle (where I as the user have complete control over the contents).

But I'm not convinced this leads to a better user experience. As an end-user of these functions, I can see this approach pushing some friction into my workflow. I think there's utility in retaining an `encode_shinylive_url()` that matches the API of the CLI command and that does all of the "magic" for the user. Would this be a reasonable middle ground?

Reply:
I relocated the app language detection so that we don't need a new app bundle type, we can just use `list[FileContentJson]` for the app payload.

I also added `encode_shinylive_url()`, which now serves as a programmatic sibling of the CLI command, while retaining the individual functions for each step.